// vscode-test/out/jest-runner/debug-console/logger.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.logger = void 0;
var colorize_1 = require("./colorize");
function logger(text) {
    text = text.replace(/\n$/, '');
    var message = formatTestHeader(text);
    if (!message) {
        message = formatTestDescription(text);
    }
    if (!message) {
        message = formatTestError(text);
    }
    if (!message) {
        message = formatSnapshotMessage(text);
    }
    if (!message) {
        message = formatTestSummary(text);
    }
    if (process.env.NODE_ENV === 'test') {
        return message || text;
    }
    console.log(message || text);
    return true;
}
exports.logger = logger;
function formatTestHeader(text) {
    var filepath = text.replace(/^(PASS|FAIL)/, '').trim();
    var _a = filepath.split('/').reverse(), testFilename = _a[0], testPathParts = _a.slice(1);
    var testPath = testPathParts.reverse().join('/');
    if (text.startsWith('PASS')) {
        return colorize_1.bold(colorize_1.greenBg(colorize_1.black(' PASS '))) + " " + colorize_1.darkGray(testPath + "/") + colorize_1.bold(testFilename);
    }
    if (text.startsWith('FAIL')) {
        return colorize_1.bold(colorize_1.redBg(colorize_1.black(' FAIL '))) + " " + colorize_1.darkGray(testPath + "/") + colorize_1.bold(testFilename);
    }
    return '';
}
function formatTestDescription(text) {
    if (text.includes('✓')) {
        return "  " + colorize_1.green('✓') + " " + colorize_1.darkGray(text.replace(/✓/, '').trim());
    }
    if (text.includes('✕')) {
        return "  " + colorize_1.red('✕') + " " + colorize_1.darkGray(text.replace(/✕/, '').trim());
    }
    if (text.includes('○')) {
        return "  " + colorize_1.yellow('○') + " " + colorize_1.darkGray(text.replace(/○/, '').trim());
    }
    if (text.includes('✎')) {
        return "  " + colorize_1.purple('✎') + " " + colorize_1.darkGray(text.replace(/✎/, '').trim());
    }
    return '';
}
function formatTestError(text) {
    return text.includes('●') ? colorize_1.red(text) : '';
}
function formatSnapshotMessage(text) {
    if (text.endsWith('updated.') || text.endsWith('written.') || text.endsWith('removed.')) {
        return colorize_1.bold(colorize_1.green(text));
    }
    if (text.endsWith('obsolete.')) {
        return colorize_1.bold(colorize_1.yellow(text));
    }
    if (text.endsWith('failed.')) {
        return colorize_1.bold(colorize_1.red(text));
    }
    if (text === 'Snapshot Summary') {
        return colorize_1.bold(text);
    }
    if (text.includes('written from')) {
        return formatSnapshotSummary(text, 'written', colorize_1.green);
    }
    if (text.includes('updated from')) {
        return formatSnapshotSummary(text, 'updated', colorize_1.green);
    }
    if (text.includes('removed from')) {
        // Use custom messaging for removed snapshot files
        if (text.includes('file')) {
            var _a = /(\d)+/.exec(text), numSnapshots = _a[0], numTestSuites = _a[1];
            return " " + colorize_1.bold(colorize_1.green("\u203A " + numSnapshots + " snapshot " + (Number(numSnapshots) > 1 ? 'files' : 'file') + " removed")) + " from " + numTestSuites + " " + (Number(numTestSuites) > 1 ? 'test suites' : 'test suite') + ".";
        }
        return formatSnapshotSummary(text, 'removed', colorize_1.green);
    }
    if (text.includes('obsolete from')) {
        return formatSnapshotSummary(text, 'obsolete', colorize_1.yellow) + " " + colorize_1.darkGray('To remove them all, re-run jest with `JEST_RUNNER_UPDATE_SNAPSHOTS=true`.');
    }
    if (text.includes('↳')) {
        var filepath = text.replace(/↳/, '').trim();
        var _b = filepath.split('/').reverse(), testFilename = _b[0], testPathParts = _b.slice(1);
        var testPath = testPathParts.reverse().join('/');
        return "   \u21B3 " + colorize_1.darkGray(testPath + "/") + colorize_1.bold(testFilename);
    }
    if (text.includes('failed from')) {
        return formatSnapshotSummary(text, 'failed', colorize_1.red) + " " + colorize_1.darkGray('Inspect your code changes or re-run jest with `JEST_RUNNER_UPDATE_SNAPSHOTS=true` to update them.');
    }
    return '';
}
function formatSnapshotSummary(text, status, colorFunc) {
    var _a = /(\d)+/.exec(text), numSnapshots = _a[0], numTestSuites = _a[1];
    return " " + colorize_1.bold(colorFunc("\u203A " + numSnapshots + " " + (Number(numSnapshots) > 1 ? 'snapshots' : 'snapshot') + " " + status)) + " from " + numTestSuites + " " + (Number(numTestSuites) > 1 ? 'test suites' : 'test suite') + ".";
}
function formatTestSummary(text) {
    if (!text.includes('\n')) {
        return '';
    }
    var summary = [];
    for (var _i = 0, _a = text.split('\n'); _i < _a.length; _i++) {
        var line = _a[_i];
        if (line.includes('Ran all test suites.')) {
            summary.push(colorize_1.darkGray(line));
            continue;
        }
        if (line.includes('Test Suites:')) {
            line = line.replace('Test Suites:', colorize_1.bold('Test Suites:'));
        }
        if (line.includes('Tests:')) {
            line = line.replace('Tests:', colorize_1.bold('Tests:'));
        }
        if (line.includes('Snapshots:')) {
            line = line.replace('Snapshots:', colorize_1.bold('Snapshots:'));
        }
        if (line.includes('Time:')) {
            line = line.replace('Time:', colorize_1.bold('Time:'));
        }
        if (line.includes('passed')) {
            line = line.replace(/(?<num>\d+) passed/, colorize_1.bold(colorize_1.green('$<num> passed')));
        }
        if (line.includes('updated')) {
            line = line.replace(/(?<num>\d+) updated/, colorize_1.bold(colorize_1.green('$<num> updated')));
        }
        if (line.includes('written')) {
            line = line.replace(/(?<num>\d+) written/, colorize_1.bold(colorize_1.green('$<num> written')));
        }
        if (line.includes('removed')) {
            // Use custom messaging for removed snapshot files
            line = line.replace(/(?<num>\d+) (?<fileText>file|files) removed/, colorize_1.bold(colorize_1.green('$<num> $<fileText> removed')));
            line = line.replace(/(?<num>\d+) removed/, colorize_1.bold(colorize_1.green('$<num> removed')));
        }
        if (line.includes('todo')) {
            line = line.replace(/(?<num>\d+) todo/, colorize_1.bold(colorize_1.purple('$<num> todo')));
        }
        if (line.includes('skipped')) {
            line = line.replace(/(?<num>\d+) skipped/, colorize_1.bold(colorize_1.yellow('$<num> skipped')));
        }
        if (line.includes('obsolete')) {
            line = line.replace(/(?<num>\d+) obsolete/, colorize_1.bold(colorize_1.yellow('$<num> obsolete')));
        }
        if (line.includes('failed')) {
            line = line.replace(/(?<num>\d+) failed/, colorize_1.bold(colorize_1.red('$<num> failed')));
        }
        summary.push(line);
    }
    return summary.join('\n');
}
